lib/records/RecProtoTags.h

```diff
@@ -0,0 +1,79 @@
+/** @file
+
+  Protocol Tags
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#ifndef _REC_PROTO_TAGS_H_
+#define _REC_PROTO_TAGS_H_
+
+#include <unordered_set>
+#include "ts/ink_inet.h"
+#include "ts/Map.h"
+
+/* Protocol session well-known protocol names.
+   These are also used for NPN setup.
+*/
+
+const char *const TS_ALPN_PROTOCOL_HTTP_0_9 = IP_PROTO_TAG_HTTP_0_9.ptr();
+const char *const TS_ALPN_PROTOCOL_HTTP_1_0 = IP_PROTO_TAG_HTTP_1_0.ptr();
+const char *const TS_ALPN_PROTOCOL_HTTP_1_1 = IP_PROTO_TAG_HTTP_1_1.ptr();
+const char *const TS_ALPN_PROTOCOL_HTTP_2_0 = IP_PROTO_TAG_HTTP_2_0.ptr();
+
+const char *const TS_PROTO_TAG_HTTP_1_0 = TS_ALPN_PROTOCOL_HTTP_1_0;
+const char *const TS_PROTO_TAG_HTTP_1_1 = TS_ALPN_PROTOCOL_HTTP_1_1;
+const char *const TS_PROTO_TAG_HTTP_2_0 = TS_ALPN_PROTOCOL_HTTP_2_0;
+const char *const TS_PROTO_TAG_TLS_1_3  = IP_PROTO_TAG_TLS_1_3.ptr();
+const char *const TS_PROTO_TAG_TLS_1_2  = IP_PROTO_TAG_TLS_1_2.ptr();
+const char *const TS_PROTO_TAG_TLS_1_1  = IP_PROTO_TAG_TLS_1_1.ptr();
+const char *const TS_PROTO_TAG_TLS_1_0  = IP_PROTO_TAG_TLS_1_0.ptr();
+const char *const TS_PROTO_TAG_TCP      = IP_PROTO_TAG_TCP.ptr();
+const char *const TS_PROTO_TAG_UDP      = IP_PROTO_TAG_UDP.ptr();
+const char *const TS_PROTO_TAG_IPV4     = IP_PROTO_TAG_IPV4.ptr();
+const char *const TS_PROTO_TAG_IPV6     = IP_PROTO_TAG_IPV6.ptr();
+
+struct tag_pred {
+  bool
+  operator()(char const *x, char const *y) const
+  {
+    return strncmp(x, y, strlen(x)) == 0;
+  }
+};
+
+struct tag_hash {
+  size_t
+  operator()(char const *x) const
+  {
+    size_t h = 0;
+    while (*x) {
+      h = h * 27 + (unsigned char)*x++;
+    }
+    return h;
+  }
+};
+
+typedef std::unordered_set<char const *, tag_hash, tag_pred> TSProtoTagsMap;
```

Comment:
We can't use one of our existing hash maps for this?

Reply:
We can and my initial implementation did, but as I mentioned in the comments I found it very difficult to isolate the code in that case so that it was testable. I went down a rabbit hole of stubbing out symbols from all over the place.

I'm willing to go back to using the internal map, but we will need to do some larger refactoring if we want effective tests I think. I'm open to suggestions.

Reply:
What are the problems with testing with the internal map?  Please give details.  I would like to resolve this issue instead of switching to use unordered_set.

Reply:
OK, give me some time, and I'll post details.

Reply:
I'm unclear why the internal maps are superior to `unordered_set` in this situation. You're micro-optimizing operations that only occur at process start in exchange for using non-standard and undocumented classes.